### PR TITLE
Implement callstack level for debugger commands

### PR DESCRIPTION
**Pull request by fatcerberus**
Implement an optional callstack level for debugger commands dealing with the current function activation (such as GetLocals and Eval), allowing e.g. retrieval of variables from earlier activations and not just the topmost one.  This is especially useful when paused at an uncaught error, to ascertain the underlying cause of the crash by walking the stack.

This is implemented in a clean, backwards-compatible manner such that the debugger protocol version remains at v1.

Checklist:
- [x] Implement callstack level for GetLocals
- [x] Implement callstack level for GetVar/PutVar
- [x] Implement callstack level for Eval
- [x] Document the new parameter in `debugger.rst`
- [x] Manual testing to ensure correct and compatible functionality
- [x] Await results of branch review and clean up as necessary

### CODE REVIEW

**Comment by fatcerberus** [comment]
Regarding the FIXME: There doesn't seem to be any way to read ahead in the debug stream?

I'd like to implement this so that the callstack level is optional (defaulting to -1) to maintain protocol compatibility, but I can't risk reading ahead because if I happen to retire the EOM, then best I can tell the subsequent skip_to_eom() would skip the entire following message.

**Comment by svaarala** [comment]
Yeah, there isn't a read window or an actual "peek for byte(s)" primitive now, the "peek" is for availability of bytes only. This would need some changes to make it easier to do conditional parsing.

For this change specifically it'd actually be enough to just consume the EOM and then somehow flag that it was already consumed so that it didn't need to be scanned for again. But this'd be quite awkward, and wouldn't work in other cases.

Adding the ability to push back a limited number of bytes should work and be relatively clean. We'd then just detect those pushed back bytes in `duk_debug_read_bytes()` (before the actual transport read loop) and `duk_debug_read_peek()` (returning > 0 if there were pushed back bytes).

A limited pushback window won't be enough to handle a generic "unconsume dvalue" case because dvalues have no upper size limit - but it would be enough to unconsume an EOM specifically. It would also allow a conditional dvalue read, e.g. "read dvalue unless EOM" or "read integer dvalue (or fail if not integer)" because one byte is enough to detect the EOM. For these two conditional reads a 1-byte pushback would be enough.

We could also use a read-ahead window so that we'd always try to keep more data available than was consumed. One would have to be careful with blocking here though, as we can't always guarantee there's e.g. at least one extra byte waiting if none are available. So a limited pushback might be easier to implement?

**Comment by fatcerberus** [comment]
The EOM pushback would probably be sufficient in general for the foreseeable future.  This is really the only use case I imagine for it, to ensure `duk__debug_skip_to_eom()` doesn't eat the following message.  For any other dvalue type, if the bytes are consumed then it's a pretty safe assumption that the debug command needed them, otherwise it wouldn't have initiated the read.  It's just that if that read comes back with an EOM instead of a parameter, we need to put it back.

Another issue I can see is that there is apparently no generic helper for reading dvalues.  Concretely in this case, the `duk_debug_read_int()` call will break the connection if it encounters an EOM instead of the expected integer.  Not sure how to go about fixing that yet, I'll have to give the debugger code another lookover.

**Comment by svaarala** [comment]
Right, if peeking for EOM is the only case for the foreseeable future, then just flagging that "oops, EOM is already eaten" and checking for that in the message loop would be sufficient. Hopefully there'd be some clean way to do that.

There's a generic `duk_debug_read_tval()` which will push the value read on the value stack - you can use that for a type independent dvalue read. I wanted to avoid creating yet another intermediate tag/value representation in the debugger code so that's why most read calls will read a specific type, and the few places where an arbitrary type is accepted `duk_debug_read_tval()` is used. It won't accept EOMs etc however, because it's not a *dvalue* reader but a `duk_tval` reader.

The current deubgger implementation is a reflection of the basic idea that Duktape wouldn't accumulate too much version dependent and optional backwards compatible features. Rather, the debug protocol version would be bumped (and announced in the protocol), and the debug client would then know what Duktape expects. That's why the only way to extend messages is by appending new values (which is not very well supported by the current reading code as you noted); there's no structured value support etc.

I won't go into it too much, but my hunch is that at the moment fork lift upgrading the debugger rather doing a lot of incremental development to incorporate features like heap walking would be a better option. For backwards compatibility it'd be relatively simple to support both protocol version 1 (current) and 2 (next) at the same time, with separate `duk_debugger.c` files providing the debugger implementation. Even going forward supporting "current" and "previous" fork lift version might be reasonable, because the debugger integrates quite narrowly to other internals -- at least for the time being.

**Comment by svaarala** [comment]
All that said, I won't object to this pull if it can be implemented cleanly :-)

**Comment by fatcerberus** [comment]
I'm generally good at maintaining compatibility in a clean way; I managed to maintain full Sphere 1.5 compatibility in minisphere 2.0 without any ugly hacks, so I'll see what I can do here.

My main thing is I don't want to bump the protocol version unnecessarily for such a tiny change, and it probably makes sense for the call index to be optional anyway (using the top activation by default), even for clients which are aware of the extra parameter.

**Comment by svaarala** [comment]
It's not so much the difficulty but the effort related in updating documentation incrementally, regression testing, etc. Sometimes this is unavoidable, and it does affect other areas of Duktape.

Sometimes it's cleaner and more efficient to do "feature packs" rather than individual features. With the debugger this might be the case, because the current debugger version really is just a "first pass" at debugger support. It was quite tightly timeboxed too, so several compromises had to be made. To improve on the basic structure, backwards incompatible changes will be needed pretty soon, and I hope to avoid a too long backwards compatibility baggage going forwards.

Obviously this is balancing between user convenience and overall project progress. That's why I've been considering support a relatively static "previous" (or "stable") debugger protocol version and a "current" (or "in progress") debugger protocol version simultaneously. Incompatible changes could be made freely for the "in progress" version while the "stable" version wouldn't gain much features but we wouldn't be causing debug clients too much hassle until they're ready to upgrade. At any point we'd only support two protocol versions so old features would phase out naturally.

To be clear, I don't mean the breaking changes will be "rewrite 50% of the debug client" :) But rather, we wouldn't need to hesitate about reorganizing some command format if it needs to be done. Longer term this might be more desirable than shoehorning additional parameters in, or establishing new commands with slight parameter variations (for example), because all that stuff adds up over time.

The other approach would be for the protocol to be more easily extensible (support structured values with optional keys for example instead of relying on a list of dvalues only). But that has other costs, of course.

Anyway, this was more of a high level issue, not really an issue for this pull.

**Comment by fatcerberus** [comment]
I figured out a way to do this cleanly and generically:

1. Implement a one-byte pushback, like C `ungetc()`.
2. Add a helper to read the next byte from the stream, and immediately push it back using the aforementioned mechanism.  Conceptually this is simply a one-byte lookahead.
3. Based on the byte returned to the caller, it can decide how to proceed.  Since dvalue type is always encoded by a single byte and Duktape assumes it can read an entire message at a time, unnecessary blocking isn't an issue.
4. In any case, the terminating EOM is never accidentally eaten (because.Step 2 puts it back).

**Comment by svaarala** [comment]
That sounds good to me -- the pushback byte would then need special handling in the lowest level read primitive but nothing above that (including "skip bytes" which just reads and will just work).

**Comment by fatcerberus** [comment]
This is one of the reasons I like NUL-terminated strings (and terminators in general) too: If you can assume the terminator is always present, then you can always safely look one element ahead.

**Comment by fatcerberus** [comment]
Quick question that I can't ascertain from looking over the debugger code: Is `duk_debug_read_bytes()` guaranteed to be called with `length >= 1`?  Or can the length be zero in the case of, e.g. empty strings?

**Comment by fatcerberus** [comment]
@svaarala Here's a much better implementation.  Let me know what you think.

**Comment by svaarala** [comment]
A zero-read is possible yes, at least for strings. It might be good to note that in a comment because as you say it's not apparent.

**Comment by svaarala** [comment]
Looks pretty good to me at a glance :) Adding initialization/assert for at least "have next byte" is needed to handle multiple attaches.

**Comment by fatcerberus** [comment]
Calling attach while already attached is a no-op right?  `heap->dbg_have_next_byte` is set to 0 (false) on detach, so an assert should be sufficient I think.

**Comment by svaarala** [comment]
Hmm, I guess attach should actually check for an attach without previous detach - but it doesn't now.

Is there a guarantee `dbg_have_next_byte != 0` if, say, an error is thrown during debugger message handling and a followup detach?

I guess the consistent handling would be to zero them out in `duk_debug_do_detach()` and initialize them in `duk_debugger_attach()` as is done for other variables. That pushback state is associated with a specific transport so it should never carry over a transport change.

**Comment by fatcerberus** [comment]
`duk_debug_do_detach()` in the current branch sets `dbg_have_next_byte` to zero, so as long as something doesn't go horribly wrong causing the detach handler not to be called, that part should be fine.

I agree that it should be explicitly initialized to zero on attach, I'll add that.

**Comment by fatcerberus** [comment]
Somewhat related: I tried to implement a callstack level for Eval as well, but ran into an issue: the Eval handler simply calls the built-in `eval()` which naturally always runs in the context of the current activation.  I first considered a hack like this:

```javascript
	old_top = thr->callstack_top;
	thr->callstack_top -= (level + 1);
	call_ret = duk_handle_call(thr, 1 /*num_stack_args*/, call_flags);
	thr->callstack_top = old_top;
```

But then realized that would clobber the activations above `level`.  So it doesn't seem like implementing this for Eval will be an easy change.  I'll look into it further after this is merged.

**Comment by fatcerberus** [comment]
Alright, I cleaned up the changes, added some comments, and also implemented callstack level for GetVar while I was at it.

**Comment by fatcerberus** [comment]
By the way, I managed to get callstack level working for Eval in my local copy, as well. :)  As I said, I'll open a separate pull for that once everything has been sorted out in this one.

**Comment by svaarala** [comment]
Setting the top would also change GC reachability which would probably be unfixable.

The easiest change is probably to refactor eval() a little bit - basically you want to compile a function (eval code) and give it an explicit environment record. I can help with that change - maybe we could do it in a separate pull?

**Comment by fatcerberus** [comment]
I managed to get it working in my local copy by passing a callstack index as a second argument to the built-in Duktape/C eval function.  Ecmascript evals never see this because the built-in function is pushed with nargs == 1. :)

I didnt add it to this pull though because I think it's probably best to do the Eval change in a separate pull, there are a lot more things that can go wrong. JS eval is a very strange beast.

By the way, how do I add the mergeready label?  It doesn't seem to let me edit the labels for the pull.

**Comment by fatcerberus** [comment]
There may be a bug in this related to running GetLocals/GetVar against a native activation (Duktape doesn't send a reply), further testing will be needed to ensure the bug is in Duktape code and not my client implementation.

**Comment by fatcerberus** [comment]
If you're curious, here's a prototype I came up with for Eval callstack level:
https://github.com/fatcerberus/duktape/commit/a62a146e5cdbf3882af08ea0719f2ea01870b051

**Comment by fatcerberus** [comment]
Follow up to above: The implementation works flawlessly, it was my glitchy thrown-together client. :)

I consider this merge-ready from my end.  We'll do the Eval change in a separate pull, the current setup I have for that is kind of hacky.

**Comment by svaarala** [comment]
@fatcerberus Sounds good :) I've been quite busy the past few days (and will be for a few more), I'll try to look into this ASAP. As for the labeling, I guess you can't enable public labeling which is too bad :/

File: src/duk_debugger.c [hunk 1]
```diff
@@ -149,6 +150,22 @@ DUK_INTERNAL void duk_debug_skip_byte(duk_hthread *thr) {
  *  Debug connection read primitives
  */
 
+/* Peek ahead in the stream one byte. */
+DUK_INTERNAL uint8_t duk_debug_peek_byte(duk_hthread *thr)
+{
```
Comment (svaarala) [inline src/duk_debugger.c]: Style trivia: opening brace on same line as function args.

**Comment by svaarala** [comment]
After a first read-through of the diff, everything seemed logical. That pushback byte worked its way in there pretty cleanly after all :) I'll give it a bit more thorough read-through tomorrow and see if I could add some support into the web UI (useful for manual regression testing too).

**Comment by fatcerberus** [comment]
Okay, I fixed the style issue you mentioned and also implemented callstack level for PutVar as well, for symmetry with GetVar.

**Comment by fatcerberus** [comment]
After quite a bit of real-world testing I changed my mind: I cleaned up the Eval changes and merged them into this branch.  I still consider this merge-ready.

File: src/duk_debugger.c [hunk 2]
```diff
@@ -1284,27 +1346,41 @@ DUK_LOCAL void duk__debug_handle_eval(duk_hthread *thr, duk_heap *heap) {
...
+	/* nargs == 2 so we can pass a callstack level to eval(). */
+	duk_push_c_function(ctx, duk_bi_global_object_eval, 2 /*nargs*/);
 	duk_push_undefined(ctx);  /* 'this' binding shouldn't matter here */
 	(void) duk_debug_read_hstring(thr);
+	if (duk_debug_peek_byte(thr) != 0x00) {
+		level = duk_debug_read_int(thr);  /* optional callstack level */
+		if (level >= 0 || -level > (duk_int32_t)thr->callstack_top) {
+			DUK_D(DUK_DPRINT("invalid callstack level for Eval"));
+			duk_debug_write_error_eom(thr, DUK_DBG_ERR_NOTFOUND, "invalid callstack level");
+			return;
```
Comment (svaarala) [inline src/duk_debugger.c]: When this error happens the current code leaves temporaries on the value stack. These should cause an assert failure in the debugger "process messages" loop.

It might actually be less error prone to change the "process messages" loop to store the top-at-entry value, and forcibly reset the value stack top when completing a message handle: this would be more robust and would allow error paths to just return instead of carefully pop stuff first.

Comment (fatcerberus) [inline src/duk_debugger.c]: So in other words, the same way Duktape handles unmatched pushes in a Duktape/C function.  That's probably a good idea, it will make error handling cleaner at least.  That said, I'll fix the bug here. :-)

**Comment by svaarala** [comment]
Re: that value stack thing, the best place might be in "duk__debug_process_message".

I'll merge a cleanup step into master with this change - I'll also clean the other message handlers.

**Comment by fatcerberus** [comment]
So do you want me to fix the bug here, or...?

**Comment by svaarala** [comment]
It's probably best to make the structure more robust - I'll post a branch shortly with the message loop change. If you rebase on that there should be no need to fix the issue.

**Comment by svaarala** [comment]
Ok, this is the pull: https://github.com/svaarala/duktape/pull/419

I'll test the debugger functionality before merging, should be in master soon.

**Comment by svaarala** [comment]
Ok that value stack change is in master now.

**Comment by fatcerberus** [comment]
Thanks, that was quick. :-)  Give me a minute to fix the conflict and make sure everything's still correct and then I'll push the rebased branch.

**Comment by fatcerberus** [comment]
Okay, rebased on top of latest master, waiting on Travis checks.

**Comment by fatcerberus** [comment]
With regard to the protocol version discussion earlier, is that versioned independently of Duktape itself?  e.g. Would be considered acceptable in general to bump the protocol version in a minor release?

**Comment by svaarala** [comment]
Versioning is still a bit of a work in progress. The explicit version is there so that the wire protocol can be identified directly without having a Duktape version table or similar in the debug client which is good for forwards compatibility (how would the debug client know how future Duktape versions map to protocol versions).

Having a separate version also makes it possible for a certain Duktape version to support two debugger protocol versions (selectable through config options) if that turns out to be a good approach - which means that a Duktape version isn't necessarily 1:1 with the debug protocol version.

For now, I think it'd be best to make incompatible debugger protocol changes in Duktape major version bumps only.

**Comment by fatcerberus** [comment]
One thing that probably needs cleanup is the documentation for the Eval command.  I wasn't sure how to properly reword it to reflect the new functionality.

**Comment by fatcerberus** [comment]
Did you need me to do anything else for this?

**Comment by svaarala** [comment]
Not off the top of my head, I'm hoping to add something to the web UI to be able to manually test this feature for regressions. Or maybe I'll just use the JSON proxy for that.

**Comment by svaarala** [comment]
Did some testing with the JSON proxy, and mostly things seem to be working fine.

However, found the following issue with Eval. The callstack here is: global -> foo() -> bar():

```
Error: aiee
	eval input:1 preventsyield
	anon  native strict directeval preventsyield
	bar _test.js:5
	foo _test.js:10
	global _test.js:13 preventsyield
```

Doing an Eval request with an arbitrary call stack index:

```
{"request":"Eval","args":["print(new Error('aiee').stack)",-2]}
{"notify":"Print","command":2,"args":["Error: aiee\n\teval input:1 preventsyield\n\tanon  native strict directeval preventsyield\n\tbar _test.js:5\n\tfoo _test.js:10\n\tglobal _test.js:13 preventsyield\n"]}
```

When the eval level is given as e.g. -2 or -3, the error stack trace will still reflect the whole call stack so that it's independent of what level is given.

After all, the call stack is in place and hasn't been dismantled for the duration of Eval, so that any internal code dealing with the call stack will still see a full call stack. Also any new calls made by the code being Eval'd will conceptually be on top of the full call stack - only the lexical environment of the Eval has been changed. A similar issue would happen for `Duktape.act()` where the indices would map to the full call stack regardless of the given Eval level.

I think this needs at least a short note in the documentation, e.g. the Eval happens in the lexical scope indicated by the call stack level index, but the call stack itself is unchanged so that e.g. tracebacks and `Duktape.act()` will see the full call stack.

**Comment by svaarala** [comment]
Here's a JSON proxy example for reading local variables off every level:

```
$ telnet localhost 9093
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
{"notify":"_Connected","args":["1 10399 v1.2.0-778-gd7838fe duk command built from Duktape repo"]}
{"notify":"Status","command":1,"args":[1,"_test.js","global",1,0]}
{"request":"AddBreak","args":["_test.js", 7]}
{"reply":true,"args":[0]}
{"request":"Resume"}
{"reply":true,"args":[]}
{"notify":"Status","command":1,"args":[0,"_test.js","global",1,0]}
{"notify":"Status","command":1,"args":[1,"_test.js","bar",7,1]}
{"request":"GetLocals","args":[]}
{"reply":true,"args":["b",{"type":"number","data":"401c000000000000"},"y",{"type":"number","data":"4000000000000000"}]}
{"request":"GetLocals","args":[-1]}
{"reply":true,"args":["b",{"type":"number","data":"401c000000000000"},"y",{"type":"number","data":"4000000000000000"}]}
{"request":"GetLocals","args":[-2]}
{"reply":true,"args":["a",{"type":"number","data":"4059000000000000"},"bar",{"type":"object","class":6,"pointer":"000000000260c9e0"},"x",{"type":"number","data":"3ff0000000000000"}]}
{"request":"GetLocals","args":[-3]}
{"reply":true,"args":[]}
{"request":"GetLocals","args":[-4]}
{"error":true,"args":[3,"invalid callstack level"]}
```

File: src/duk_bi_global.c [hunk 3]
```diff
@@ -448,15 +449,27 @@ DUK_INTERNAL duk_ret_t duk_bi_global_object_eval(duk_context *ctx) {
 		return 1;  /* return arg as-is */
 	}
 
+#if defined(DUK_USE_DEBUGGER_SUPPORT)
+	/*  XXX: level is used only by the debugger and should never be present
+	 *  for eval() called from Ecmascript code. is there a way to assert for
+	 *  this?
+	 */
```
Comment (svaarala) [inline src/duk_bi_global.c]: I think it'd be just fine to comment shortly here that level is present only for debugger-initiated calls. The actual eval() binding has nargs==1 but it's not easy to check here.

**Comment by fatcerberus** [comment]
The callstack behavior mentioned above seems intuitive to me, at least.  Agreed that a note in the documentation should be sufficient.

File: src/duk_debugger.c [hunk 4]
```diff
@@ -1113,16 +1139,27 @@ DUK_LOCAL void duk__debug_handle_get_var(duk_hthread *thr, duk_heap *heap) {
...
 	duk_bool_t rc;
+	duk_int32_t level;
 
 	DUK_UNREF(heap);
 	DUK_D(DUK_DPRINT("debug command GetVar"));
 
 	str = duk_debug_read_hstring(thr);  /* push to stack */
 	DUK_ASSERT(str != NULL);
+	if (duk_debug_peek_byte(thr) != 0x00) {
+		level = duk_debug_read_int(thr);  /* optional callstack level */
```
Comment (svaarala) [inline src/duk_debugger.c]: Here 0x00 could be `DUK_DBG_MARKER_EOM`.

File: src/duk_bi_global.c (hunk 3 above)
Comment (fatcerberus) [inline src/duk_bi_global.c]: I always thought nargs for Duktape/C functions was pointless (I just use `DUK_VARARGS` for everything), but it came in really handy here. :)

File: src/duk_debugger.c [hunk 5]
```diff
@@ -1159,10 +1197,20 @@ DUK_LOCAL void duk__debug_handle_put_var(duk_hthread *thr, duk_heap *heap) {
 	duk_debug_read_tval(thr);           /* push to stack */
 	tv = duk_get_tval(ctx, -1);
 	DUK_ASSERT(tv != NULL);
+	if (duk_debug_peek_byte(thr) != 0x00) {
+		level = duk_debug_read_int(thr);  /* optional callstack level */
```
Comment (svaarala) [inline src/duk_debugger.c]: 0x00 -> `DUK_DBG_MARKER_EOM`.

File: src/duk_debugger.c [hunk 6]
```diff
@@ -1229,15 +1277,28 @@ DUK_LOCAL void duk__debug_handle_get_call_stack(duk_hthread *thr, duk_heap *heap
...
 	duk_activation *curr_act;
+	duk_int32_t level;
 	duk_hstring *varname;
 
 	DUK_UNREF(heap);
 
-	duk_debug_write_reply(thr);
-	if (thr->callstack_top == 0) {
-		goto callstack_empty;
+	if (duk_debug_peek_byte(thr) != 0x00) {
```
Comment (svaarala) [inline src/duk_debugger.c]: 0x00 -> `DUK_DBG_MARKER_EOM`.

File: src/duk_debugger.c [hunk 7]
```diff
@@ -1272,28 +1333,44 @@ DUK_LOCAL void duk__debug_handle_eval(duk_hthread *thr, duk_heap *heap) {
...
+	 * a 'direct call to eval'.
 	 */
 
-	duk_push_c_function(ctx, duk_bi_global_object_eval, 1 /*nargs*/);
+	/* nargs == 2 so we can pass a callstack level to eval(). */
+	duk_push_c_function(ctx, duk_bi_global_object_eval, 2 /*nargs*/);
 	duk_push_undefined(ctx);  /* 'this' binding shouldn't matter here */
+
 	(void) duk_debug_read_hstring(thr);
+	if (duk_debug_peek_byte(thr) != 0x00) {
```
Comment (svaarala) [inline src/duk_debugger.c]: 0x00 -> `DUK_DBG_MARKER_EOM`.

File: src/duk_debugger.c [hunk 8]
```diff
@@ -1272,28 +1333,44 @@ DUK_LOCAL void duk__debug_handle_eval(duk_hthread *thr, duk_heap *heap) {
...
 
 	DUK_D(DUK_DPRINT("debug command Eval"));
 
-	/* The eval code must be executed within the current (topmost)
-	 * activation.  For now, use global object eval() function, with
-	 * the eval considered a 'direct call to eval'.
+	/* The eval code must be executed within the specified activation.
+	 * For now, use global object eval() function, with the eval considered
+	 * a 'direct call to eval'.
 	 */
```
Comment (svaarala) [inline src/duk_debugger.c]: Perhaps note here, too, that the callstack index affects the scope seen by Eval but not the call stack (i.e. full call stack will be visible in tracebacks etc).

**Comment by svaarala** [comment]
All right, the trivia is commented in commits and the other stuff I had was:

- Comment on the Eval call stack behavior briefly in the debugger document.
- Remove call stack level references from debugger future work as appropriate (there are a few clear references to this feature)

Good stuff :)

File: src/duk_bi_global.c (hunk 3 above)
Comment (svaarala) [inline src/duk_bi_global.c]: Yeah, it's intended for situations where the function really just wants a fixed configuration in its entry value stack and doesn't want to worry about missing/extra arguments. Both are useful I guess :)

**Comment by svaarala** [comment]
Action points for separate pulls (for me) are:

- Rebase the reattach fix branch and ensure `dbg_have_next_byte` and `dbg_next_byte` are initialized properly.
- Work in minimal callstack level support into the web UI somehow. Perhaps so that if you click on a callstack level, it will be used in followup Eval/GetVar/PutVar calls.

File: src/duk_debugger.c (hunk 7 above)
Comment (fatcerberus) [inline src/duk_debugger.c]: See, this was my first instinct too, but then my OCD got the better of me (because the `peek`'d byte isn't necessary always a dvalue marker, although it is in this case).

Comment (svaarala) [inline src/duk_debugger.c]: That'd be true in the general case, but isn't it necessarily an "initial byte" here (and markers are initial bytes)?

Comment (svaarala) [inline src/duk_debugger.c]: I have a little bit of OCD myself when it comes to code trivia, so I can symphatize on the hard but small choices :-)

Comment (fatcerberus) [inline src/duk_debugger.c]: Yes, in this case it would always be an IB.  I just tend to think in the general case "by default".  Anyway I'll fix these along with the other issues you brought up.

**Comment by fatcerberus** [comment]
Regarding the debugger UI, how I implemented it in the minisphere debugger is if you double-click a callstack entry, it updates the local variable list by calling GetLocals at that level, and then subsequent Eval commands operate at that level.

**Comment by svaarala** [comment]
Sounds reasonable - I'd need to maybe add highlighting for the current level, etc. The web UI was written pretty much as a throw-away prototype though so I'm not sure how far I can take it without improving the structure a bit first :)

**Comment by fatcerberus** [comment]
Okay, rebased and trivia addressed.  How's this?

File: src/duk_debugger.c [hunk 9]
```diff
@@ -149,6 +150,21 @@ DUK_INTERNAL void duk_debug_skip_byte(duk_hthread *thr) {
...
 
+/* Peek ahead in the stream one byte. */
+DUK_INTERNAL uint8_t duk_debug_peek_byte(duk_hthread *thr) {
+	/* It is important not to call this if the last byte read was an EOM.
+	 * Reading ahead in this scenario would cause unnecessary blocking if
+	 * another message is not available.
+	 */
+
+	duk_uint8_t x;
+
```
Comment (svaarala) [inline src/duk_debugger.c]: By the way, should there be a `DUK_ASSERT(thr->heap->dbg_have_next_byte == 0);` here? In other words, it's not allowed to call `duk_debug_peek_byte()` twice without consuming the pushed back byte first?

**Comment by svaarala** [comment]
One more commit note on an issue I just spotted, and a short note on Eval call stack behavior on `debugger.rst` would be nice. But otherwise good to go :)

File: src/duk_debugger.c (hunk 9 above)
Comment (fatcerberus) [inline src/duk_debugger.c]: Actually that's safe--it will just read back the pushback byte. :)

Comment (svaarala) [inline src/duk_debugger.c]: Ah true, never mind :)

**Comment by fatcerberus** [comment]
Ah, forgot to add a comment on that in debugger.rst, sorry about that.  I'll do that after my dinner. :)

**Comment by fatcerberus** [comment]
Okay, that's done.

**Comment by svaarala** [comment]
Ok, is this merge ready or any incoming changes?

**Comment by fatcerberus** [comment]
It's merge ready as long as toy have no further issues.

**Comment by svaarala** [comment]
Ok, I'll merge this and start a separate pull for the web UI stuff etc. Thanks for the nice work :)

**Comment by fatcerberus** [comment]
"Toy" should be "you", my phone is weird. :)
